File: basic_dsa_three_hundred_sixty_five.java
/**
 * Basic DSA Problems - Water and Jug Problem (LeetCode 365)
 * Can we measure target using two jugs (fill, empty, pour)?
 */
public class basic_dsa_three_hundred_sixty_five {

    // Problem: jug1, jug2 capacity. Measure exactly target? Operations: fill, empty, pour.

    /**
     * Target must be <= jug1+jug2 and divisible by gcd(jug1, jug2).
     * Bézout: ax + by = g. Any multiple of g is achievable.
     * Time O(log(min(a,b)))
     */
    public static boolean canMeasureWater(int jug1Capacity, int jug2Capacity, int targetCapacity) {
        if (targetCapacity > jug1Capacity + jug2Capacity) return false;
        return targetCapacity % gcd(jug1Capacity, jug2Capacity) == 0;
    }

    private static int gcd(int a, int b) {
        return b == 0 ? a : gcd(b, a % b);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Water and Jug Problem");
        System.out.println("=".repeat(60));

        System.out.println("3,5,4 -> " + canMeasureWater(3, 5, 4));
        System.out.println("2,6,5 -> " + canMeasureWater(2, 6, 5));
        System.out.println("1,2,3 -> " + canMeasureWater(1, 2, 3));
    }
}

File: basic_dsa_three_hundred_sixty_five.cpp
/**
 * Basic DSA - Target Sum (LeetCode 494)
 * Ways to assign +/- to nums so sum equals target.
 */
#include <iostream>
#include <string>
#include <vector>
#include <numeric>

int findTargetSumWays(std::vector<int>& nums, int target) {
    int s = std::accumulate(nums.begin(), nums.end(), 0);
    if ((s + target) % 2 || target > s || target < -s) return 0;
    int t = (s + target) / 2;
    std::vector<int> dp(t + 1);
    dp[0] = 1;
    for (int x : nums)
        for (int j = t; j >= x; j--)
            dp[j] += dp[j - x];
    return dp[t];
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Target Sum\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {1, 1, 1, 1, 1};
    std::cout << "[1,1,1,1,1], 3 -> " << findTargetSumWays(v1, 3) << "\n";

    std::vector<int> v2 = {1};
    std::cout << "[1], 1 -> " << findTargetSumWays(v2, 1) << "\n";
}
